src/httpbeast.nim

```diff
@@ -83,18 +87,30 @@ template handleClientClosure(selector: Selector[Data],
                              fd: SocketHandle|int,
                              inLoop=true) =
   # TODO: Logging that the socket was closed.
-
   # TODO: Can POST body be sent with Connection: Close?
+  var data: ptr Data = addr selector.getData(fd)
+  let completed = data.completed
+  data.closed = true
+  if not completed or fd in selector:
+    selector.unregister(fd)
+  if completed:
+    fd.SocketHandle.close()
 
-  selector.unregister(fd)
-  fd.SocketHandle.close()
-  when inLoop:
-    break
-  else:
-    return
+  when inLoop: break
+  else: return
 
 proc onRequestFutureComplete(theFut: Future[void],
                              selector: Selector[Data], fd: int) =
+  var data: ptr Data = addr selector.getData(fd)
+  if data == nil:
+    fd.SocketHandle.close()
+  elif data.fdKind == Client:
+    data.completed = true
+    if not data.closed and data.sendQueue.len == 0:
+      selector.unregister(fd)
+    if data.closed or data.sendQueue.len == 0:
+      fd.SocketHandle.close()
```

Comment:
This code seems wrong to me, if I understand correctly we will close sockets after every request. But IIRC we get multiple requests on each socket.

Reply:
If the client has closed the connection, the request has completed, and there is nothing more to send, why would you keep the socket open?